diff --git a/src/client/client/client.c b/src/client/client/client.c
index f662d8f..9843591 100644
--- a/src/client/client/client.c
+++ b/src/client/client/client.c
@@ -72,7 +72,8 @@ ElfW(Addr) libc_loadoffset, interp_loadoffset;
 /* location has the realize'd path to the local file cache. orig_location is not realized and
  * may contain symlinks
  */
-char *location, *orig_location, *cachepath;
+static char *location, *orig_location;
+static char *chosen_realized_cachepath, *chosen_parsed_cachepath, *chosen_symbolic_cachepath;
 number_t number;
 static int have_stat_patches;
 
@@ -186,6 +187,7 @@ void int_spindle_test_log_msg(char *buffer)
 
 static int init_server_connection()
 {
+   char *cachepath; // QQQ Remove after send_cachepath_query() is working
    char *connection, *rankinfo_s, *opts_s, *cachesize_s;
    int old_ldcsid;
 
@@ -197,8 +199,8 @@ static int init_server_connection()
       return 0;
 
    location = getenv("LDCS_LOCATION");
-   orig_location = getenv("LDCS_ORIG_LOCATION");
-   cachepath = getenv("LDCS_CACHEPATH");
+   orig_location = getenv("LDCS_ORIG_LOCATION");    // QQQ can we remove this?
+   cachepath = getenv("LDCS_CACHEPATH"); // QQQ Remove after send_cachepath_query() is working
    number = (number_t) strtoul(getenv("LDCS_NUMBER"), NULL, 0);
    connection = getenv("LDCS_CONNECTION");
    rankinfo_s = getenv("LDCS_RANKINFO");
@@ -255,7 +257,7 @@ static int init_server_connection()
       if (ldcsid == -1)
          return -1;
 
-      send_cachepath_query( ldcsid, &cachepath );
+      send_cachepath_query( ldcsid, &chosen_realized_cachepath, &chosen_parsed_cachepath, &chosen_symbolic_cachepath );
       send_pid(ldcsid);
       send_location(ldcsid, location);
       send_rankinfo_query(ldcsid, rankinfo+0, rankinfo+1, rankinfo+2, rankinfo+3);
@@ -469,7 +471,7 @@ char *client_library_load(const char *name)
 
    char *orig_file_name = (char *) name;
    if (is_in_spindle_cache(name)) {
-      debug_printf2("Library %s is in spindle cache (%s). Translating request\n", name, location);
+      debug_printf2("Library %s is in spindle cache (%s). Translating request\n", name, chosen_realized_cachepath);
       memset(fixed_name, 0, MAX_PATH_LEN+1);
       send_orig_path_request(ldcsid, orig_file_name, fixed_name);
       orig_file_name = fixed_name;
diff --git a/src/client/client/should_intercept.c b/src/client/client/should_intercept.c
index be0d7bb..0ebb46a 100644
--- a/src/client/client/should_intercept.c
+++ b/src/client/client/should_intercept.c
@@ -29,22 +29,27 @@
 #include "spindle_debug.h"
 
 extern int relocate_spindleapi();
+static char *cachepath, *orig_cachepath;
+
+void set_intercept_cachepath( char *chosen_realized_cachepath, char *chosen_parsed_cachepath, char *chosen_symbolic_cachepath ){
+    cachepath = chosen_realized_cachepath;
+    orig_cachepath = chosen_parsed_cachepath;
+    chosen_symbolic_cachepath = chosen_symbolic_cachepath;
+}
 
-extern char *location;
-extern char *orig_location;
 
 int is_in_spindle_cache(const char *pathname)
 {
-   static int location_size = 0;
-   static int orig_location_size = 0;
-   if (!location_size) {
-      location_size = strlen(location);
+   static int cachepath_size = 0;
+   static int orig_cachepath_size = 0;
+   if (!cachepath_size) {
+      cachepath_size = strlen(cachepath);
    }
-   if (!orig_location_size) {
-      orig_location_size = strlen(orig_location);
+   if (!orig_cachepath_size) {
+      orig_cachepath_size = strlen(orig_cachepath);
    }
-   return ((strncmp(pathname, location, location_size) == 0) ||
-           (strncmp(pathname, orig_location, orig_location_size) == 0));
+   return ((strncmp(pathname, cachepath, cachepath_size) == 0) ||
+           (strncmp(pathname, orig_cachepath, orig_cachepath_size) == 0));
 }
 
 extern int is_local_prefix(const char *path, char **cached_local_prefixes);
diff --git a/src/client/client/should_intercept.h b/src/client/client/should_intercept.h
index f6a9b51..077d18f 100644
--- a/src/client/client/should_intercept.h
+++ b/src/client/client/should_intercept.h
@@ -27,6 +27,7 @@
 #define EXCL_OPEN 2
 #define ERR_CALL 3
 
+void set_intercept_cachepath( char *chosen_realized_cachepath, char *chosen_parsed_cachepath, char *chosen_symbolic_cachepath );
 int open_filter(const char *fname, int flags);
 int fopen_filter(const char *fname, const char *flags);
 int exec_filter(const char *fname);
diff --git a/src/client/client_comlib/client_api.c b/src/client/client_comlib/client_api.c
index 51acded..b49ebbe 100644
--- a/src/client/client_comlib/client_api.c
+++ b/src/client/client_comlib/client_api.c
@@ -38,7 +38,7 @@ static struct lock_t comm_lock;
 #define COMM_UNLOCK unlock(&comm_lock)
    
 
-int send_cachepath_query( int fd, char **chosen_cachepath ){
+int send_cachepath_query( int fd, char **chosen_realized_cachepath, char **chosen_parsed_cachepath, char **chosen_symbolic_cachepath ){
    ldcs_message_t message;
    char buffer[MAX_PATH_LEN+1];
    buffer[MAX_PATH_LEN] = '\0';
@@ -59,9 +59,35 @@ int send_cachepath_query( int fd, char **chosen_cachepath ){
       err_printf("Got unexpected message of type %d\n", (int) message.header.type);
       assert(0);
    }
-   *chosen_cachepath = strdup( buffer );
+   if( chosen_symbolic_cachepath ){
+       *chosen_realized_cachepath = strdup( buffer );
+   }
+
+   COMM_LOCK;
+   client_recv_msg_static(fd, &message, LDCS_READ_BLOCK);
+   COMM_UNLOCK;
+
+   if (message.header.type != LDCS_MSG_CHOSEN_CACHEPATH || message.header.len > MAX_PATH_LEN) {
+      err_printf("Got unexpected message of type %d\n", (int) message.header.type);
+      assert(0);
+   }
+   if( chosen_parsed_cachepath ){
+       *chosen_parsed_cachepath = strdup( buffer );
+   }
 
-    return 0;
+   COMM_LOCK;
+   client_recv_msg_static(fd, &message, LDCS_READ_BLOCK);
+   COMM_UNLOCK;
+
+   if (message.header.type != LDCS_MSG_CHOSEN_CACHEPATH || message.header.len > MAX_PATH_LEN) {
+      err_printf("Got unexpected message of type %d\n", (int) message.header.type);
+      assert(0);
+   }
+   if( chosen_symbolic_cachepath ){
+       *chosen_realized_cachepath = strdup( buffer );
+   }
+
+   return 0;
 }
 
 int send_file_query(int fd, char* path, int dso, char** newpath, int *errcode) {
diff --git a/src/client/client_comlib/client_api.h b/src/client/client_comlib/client_api.h
index 6cef4e8..982c4b1 100644
--- a/src/client/client_comlib/client_api.h
+++ b/src/client/client_comlib/client_api.h
@@ -42,7 +42,7 @@ int send_orig_path_request(int fd, const char *path, char *newpath);
 int send_dirlists_request(int fd, char **local_result, char **exece_result, char **to_free);
 int send_procmaps_query(int fd, int pid, char *result);
 int send_pickone_query(int fd, char *key, int *result);
-int send_cachepath_query( int fd, char **chosen_cachepath );
+int send_cachepath_query( int fd, char **chosen_symbolic_cachepath, char **chosen_parsed_cachepath, char **chosen_realized_cachepath );
 
 int get_python_prefix(int fd, char **prefix);
 
diff --git a/src/server/auditserver/ldcs_audit_server_handlers.c b/src/server/auditserver/ldcs_audit_server_handlers.c
index 4570f8d..3fcde0d 100644
--- a/src/server/auditserver/ldcs_audit_server_handlers.c
+++ b/src/server/auditserver/ldcs_audit_server_handlers.c
@@ -2964,13 +2964,16 @@ static int handle_cachepath_consensus(ldcs_process_data_t *procdata, ldcs_messag
        procdata->cachepath = procdata->location;  // FIXME better way to think about this?
     }else{
         // ldcs_audit_server_filemngt_init() does it's own realize() pass.
-        getValidCachePathByIndex( procdata->cachepath_bitidx, NULL, &procdata->cachepath, NULL);
+        getValidCachePathByIndex( procdata->cachepath_bitidx,
+                &procdata->cachepath,
+                &procdata->parsed_cachepath,
+                &procdata->symbolic_cachepath);
     }
 
     debug_printf3("Initializing file cache location %s\n", procdata->location);
     // Enable use of cachepath here.
-    ldcs_audit_server_filemngt_init(procdata->location);
-    // ldcs_audit_server_filemngt_init(procdata->cachepath);
+    // ldcs_audit_server_filemngt_init(procdata->location); // QQQ remove when cachepath works
+    ldcs_audit_server_filemngt_init(procdata->cachepath);
 
     test_printf("<internal> cachepath=%s\n", procdata->cachepath);
     return 0;
@@ -2991,12 +2994,25 @@ static int handle_chosen_cachepath_request(ldcs_process_data_t *procdata, int nc
       return 0;
 
    msg.header.type = LDCS_MSG_CHOSEN_CACHEPATH;
+
    msg.header.len = strlen(procdata->cachepath) + 1;
    msg.data = procdata->cachepath;
+   ldcs_send_msg(connid, &msg);
+   procdata->server_stat.clientmsg.cnt++;
+   procdata->server_stat.clientmsg.time += ldcs_get_time() - client->query_arrival_time;
 
+   msg.header.len = strlen(procdata->parsed_cachepath) + 1;
+   msg.data = procdata->parsed_cachepath;
    ldcs_send_msg(connid, &msg);
    procdata->server_stat.clientmsg.cnt++;
    procdata->server_stat.clientmsg.time += ldcs_get_time() - client->query_arrival_time;
+
+   msg.header.len = strlen(procdata->symbolic_cachepath) + 1;
+   msg.data = procdata->symbolic_cachepath;
+   ldcs_send_msg(connid, &msg);
+   procdata->server_stat.clientmsg.cnt++;
+   procdata->server_stat.clientmsg.time += ldcs_get_time() - client->query_arrival_time;
+
    return 0;
 }
 
diff --git a/src/server/auditserver/ldcs_audit_server_process.h b/src/server/auditserver/ldcs_audit_server_process.h
index d54e836..99f8711 100644
--- a/src/server/auditserver/ldcs_audit_server_process.h
+++ b/src/server/auditserver/ldcs_audit_server_process.h
@@ -128,6 +128,8 @@ struct ldcs_process_data_struct
   char *location;
   char *cachepaths;
   char *cachepath;
+  char *symbolic_cachepath;
+  char *parsed_cachepath;
   int64_t cachepath_bitidx;
   char *hostname;
   char *pythonprefix;
